text/0000-build-std.md

```diff
@@ -276,52 +357,1098 @@ proposed by this RFC.
+flag to be equivalent to using a modified sysroot.
+
+Without `noprelude`, rustc implicitly inserts a `extern crate $name` when using
+`--extern`. As a consequence, if a newly-built `alloc` were passed using
+`--extern alloc=alloc.rlib` then `extern crate alloc` would not be required, but
+it would be if the pre-built `alloc` could be used. This difference in how a
+crate is made available to rustc should not be observable to the user.
+
+## Why not always rebuild when the profile changes?
+[why-not-always-rebuild-when-the-profile-changes]: #why-not-always-rebuild-when-the-profile-changes
+
+Cargo's users don't currently expect that changing any part of their profile
+configuration, such as trying a different optimisation level, would trigger a
+rebuild of the standard library. For small projects, rebuilding the standard
+library could be a significant increase in the overall build time for a project.
+If `build-std = "always"` were the default, the standard library could be
+rebuilt quite frequently without much benefit. Especially as the pre-built
+standard library is built using the release profile, all debug profile builds
+would immediately trigger a rebuild of the standard library.
+
+See
+[*Why not ship a debug profile `rust-std?*][why-not-ship-a-debug-profile-rust-std]
+and
+[*Why rebuild the standard library automatically?*][why-rebuild-the-standard-library-automatically]
+
+## Why vendor the standard library's dependencies?
+[why-vendor-the-standard-librarys-dependencies]: #why-vendor-the-standard-librarys-dependencies
+
+Vendoring the standard library's dependencies has multiple advantages..
+
+- Avoid needing to support standard library dependencies in `cargo vendor`
+- Avoid needing to support standard library dependencies in `cargo fetch`
+- Re-building the standard library does not require an internet connection
+- Standard library dependency versions are fixed to those in the `Cargo.lock`
+  anyway, so the same versions would be being downloaded every time if not
+  vendored
+
+..and few disadvantages:
+
+- A larger `rust-src` component takes up more disk space and takes longer to
+  download
+- Vendored dependencies can't be updated with the latest security fixes
+
+See
+[*Why use the lockfile of the `rust-src` component?*][why-use-the-lockfile-of-the-rust-src-component]
+
+## Why not ship a debug profile `rust-std`?
+[why-not-ship-a-debug-profile-rust-std]: #why-not-ship-a-debug-profile-rust-std
+
+As the default configuration for `build-std` is `target-modifiers`, debug builds
+of the user's crate would not trigger a rebuild of the standard library and
+would use the pre-built standard library. It would only be when
+`build-std = "always"` that any debug build would first trigger a rebuild of the
+standard library.
+
+To improve the user experience in this circumstance, it could be worth shipping
+a debug profile `rust-std`, but as this is not the common case, it isn't
+proposed in this RFC.
+
```

Comment:
Might be worth calling out that some things like intrinsics rely on optimization happening otherwise they might not lower into exactly one instruction which could be counterintuitive to users.

Reply:
Added a mention of this